coderd/coderdtest/oidctest/idp.go

```diff
@@ -58,15 +59,107 @@ type deviceFlow struct {
 	granted   bool
 }
 
+type fakeIDPProtected struct {
+	mu sync.RWMutex
+
+	issuer     string
+	issuerURL  *url.URL
+	key        *rsa.PrivateKey
+	provider   ProviderJSON
+	handler    http.Handler
+	cfg        *oauth2.Config
+	fakeCoderd func(req *http.Request) (*http.Response, error)
+}
+
+func (f *fakeIDPProtected) Issuer() string {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuer
+}
+
+func (f *fakeIDPProtected) IssuerURL() *url.URL {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuerURL
+}
+
+func (f *fakeIDPProtected) PrivateKey() *rsa.PrivateKey {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.key
+}
+
+func (f *fakeIDPProtected) Provider() ProviderJSON {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.provider
+}
+
+func (f *fakeIDPProtected) Config() *oauth2.Config {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.cfg
+}
+
+func (f *fakeIDPProtected) Handler() http.Handler {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.handler
+}
+
+func (f *fakeIDPProtected) SetIssuer(issuer string) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuer = issuer
+}
+
+func (f *fakeIDPProtected) SetIssuerURL(issuerURL *url.URL) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuerURL = issuerURL
+}
+
+func (f *fakeIDPProtected) SetProvider(provider ProviderJSON) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.provider = provider
+}
+
+func (f *fakeIDPProtected) MutateConfig(fn func(cfg *oauth2.Config)) {
+	f.mu.RLock()
+	cfg := f.cfg
+	if cfg == nil {
+		cfg = &oauth2.Config{}
+	}
+	fn(cfg)
```

Comment:
Nit: This mutates the config which is a pointer, meaning the mutex only protects `f.cfg` but not the actual `&oauth2.Config{}` which could be modified and read in a racy way? I'd suggest at least making a copy of it, but did not verify if that's sufficient.

Reply:
If we take your suggestion above then we can just mutate `f.cfg` directly.

Reply:
If we're worried about races and re-entrant locking, couldn't we just do this?

```go
	f.mu.Lock()
	if cfg == nil {
		cfg = &oauth2.Config{}
	}
	cfg := *f.cfg
	f.mu.Unlock()
	fn(cfg)
	f.mu.Lock()
	f.cfg = &cfg
	f.mu.Unlock()
```

Reply:
When I make a shallow copy of `f.cfg` it causes `TestOIDCSkipIssuer` to fail. We may sadly have to live with the possibility of re-entrant locks.